File: clients/typescript/src/live-table.ts
/**
 * LiveTable: one table's Z-set state, kept current by an async reader loop (JS has no threads, so
 * unlike the Python client's dedicated reader thread this is plain non-blocking async/await --
 * see subscribeSnapshotReplay() below for how that simplifies the subscribe/snapshot race).
 *
 * Framework-free: `web/src/hooks/useTableRows.ts` is a React hook wrapping this same logic
 * inline; this class is that logic extracted so any consumer (a React hook, a Node script, a
 * plain <script> tag) can use it without React. `onChange`/the AsyncIterable are the two ways to
 * observe it; `rows` is a frozen snapshot array, never mutated in place -- a consumer holding a
 * stale reference never sees it change out from under it (same reasoning as the Python client's
 * ".df is a projection, not a mirror").
 *
 * onChange emissions are coalesced with a LEADING edge + TRAILING coalesce window (`flushMs`,
 * default 16ms -- one frame at 60Hz, the natural ceiling for a UI consumer that cannot display
 * more than one frame per 16ms anyway; `flushMs: 0` disables coalescing entirely and emits
 * synchronously per applied batch). If at least `flushMs` has elapsed since the last emit, a batch
 * is delivered immediately -- no timer, no wait -- so a lone update on an otherwise-quiet table is
 * never held back. Only a batch that lands INSIDE the window opened by the previous emit gets
 * merged into a single pending emit, fired at `lastEmit + flushMs`; further batches inside that
 * same window merge into the same pending emit, so at most one emit is ever pending. The window
 * exists at all because a firehose of tens of thousands of deltas/sec would otherwise fire one
 * callback per delta and melt the consumer -- NOT, as an earlier version of this comment claimed,
 * because this repo's own hub-driven UI does anything similar: `web/src/hooks/useTableRows.ts`
 * calls its flush on every batch with no coalescing window of its own (its 900ms timer is the
 * flash-highlight effect, an unrelated concern).
 */

import { NotReady, StreamsForgeError } from "./errors.js";
import type { Transport } from "./transport.js";
import { ZSet, type Delta, type Entry, type Row } from "./zset.js";

const DEFAULT_FLUSH_MS = 16; // one frame at 60Hz -- see the module doc comment above
const MAX_BACKOFF_MS = 15_000;

/** `touched` is the set of canonical keys (zset.ts's `canonicalKey`) that changed in the batch(es)
 * folded into this emit -- additive alongside `rows` so an existing one-arg listener keeps working
 * unchanged (JS ignores the extra argument). Resolve a key to its row with `LiveTable.row()`; a
 * key absent there was retracted, not upserted -- see that method's doc comment. */
export type ChangeListener = (rows: readonly Row[], touched: ReadonlySet<string>) => void;

function sleep(ms: number, signal: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal.aborted) return resolve();
    const t = setTimeout(resolve, ms);
    signal.addEventListener("abort", () => {
      clearTimeout(t);
      resolve();
    });
  });
}

export class LiveTable implements AsyncIterable<readonly Row[]> {
  private zset: ZSet;
  private rowsFrozen: readonly Row[] = Object.freeze([]);
  private listeners = new Set<ChangeListener>();
  private iterWaiters: Array<(v: IteratorResult<readonly Row[]>) => void> = [];
  /** Latest-wins, capacity 1 -- NOT a queue. These are state snapshots (each carries the full
   * current `rows`, not a delta), so a consumer that stops calling `next()` for a while must catch
   * up to the LATEST state on its next call, not replay every intermediate one it missed -- an
   * unbounded array here would just be a silent memory leak under a slow `for await` body. See the
   * README's "change-notification latency and backpressure" section. */
  private iterBuffer: readonly Row[] | null = null;
  private iterDone = false;

  private readyFlag = false;
  private readyWaiters: Array<{ resolve: () => void; reject: (err: Error) => void }> = [];
  private closed = false;
  private reconnectsCount = 0;
  private seqValue = 0;

  private pendingTouched: Set<string> | null = null;
  private flushTimer: ReturnType<typeof setTimeout> | null = null;
  /** -Infinity so the very first batch on a freshly-connected table always takes the leading-edge
   * immediate path -- there has never been a prior emit to be "inside the window" of. */
  private lastEmitAt = -Infinity;

  private readerAbort = new AbortController();
  private readonly readerLoopPromise: Promise<void>;

  private constructor(
    private readonly transport: Transport,
    private readonly tableName: string,
    private readonly keyFields: readonly string[] | null,
    private readonly flushMs: number,
  ) {
    this.zset = new ZSet(keyFields);
    this.readerLoopPromise = this.runReaderLoop();
  }

  /** Subscribes, snapshots, replays -- resolves once the table has filled (or rejects with
   * NotReady past `timeoutMs`). Async factory rather than a blocking constructor: JS has no
   * synchronous "wait on a background thread" primitive, and pretending otherwise would just be
   * a Promise wearing a constructor's clothes. */
  static async connect(
    transport: Transport,
    tableName: string,
    keyFields: readonly string[] | null,
    timeoutMs = 30_000,
    flushMs = DEFAULT_FLUSH_MS,
  ): Promise<LiveTable> {
    const table = new LiveTable(transport, tableName, keyFields, flushMs);
    await table.waitUntilReady(timeoutMs);
    return table;
  }

  private waitUntilReady(timeoutMs: number): Promise<void> {
    if (this.readyFlag) return Promise.resolve();
    return new Promise((resolve, reject) => {
      const entry = {
        resolve: () => {
          clearTimeout(timer);
          resolve();
        },
        reject: (err: Error) => {
          clearTimeout(timer);
          reject(err);
        },
      };
      const timer = setTimeout(() => {
        this.readyWaiters = this.readyWaiters.filter((w) => w !== entry);
        this.close();
        reject(
          new NotReady(
            `table '${this.tableName}' did not fill within ${timeoutMs}ms -- a brand-new table gets ` +
              "no backfill, so this is expected until something pushes to it",
          ),
        );
      }, timeoutMs);
      this.readyWaiters.push(entry);
    });
  }

  // ---- public surface ----

  /** Current rows, frozen -- a fresh array each time state changes, never mutated in place. */
  get rows(): readonly Row[] {
    return this.rowsFrozen;
  }

  /** Current rows plus each one's canonical key and summed weight -- delegates to the underlying
   * ZSet's own `.entries()` (see zset.ts). */
  entries(): Entry[] {
    return this.zset.entries();
  }

  /** The current row for one canonical key (e.g. one drawn from a `ChangeListener`'s `touched`
   * set), or `undefined` if that key isn't present. A touched key absent here means the tuple was
   * retracted, not upserted -- that absence IS the delete signal, since ZSet never keeps a
   * tombstone around to say so explicitly. Delegates to ZSet.get(). */
  row(key: string): Row | undefined {
    return this.zset.get(key);
  }

  get ready(): boolean {
    return this.readyFlag;
  }

  get reconnects(): number {
    return this.reconnectsCount;
  }

  get seq(): number {
    return this.seqValue;
  }

  value(col: string, keys: Record<string, unknown>): unknown {
    for (const row of this.rowsFrozen) {
      let match = true;
      for (const [k, v] of Object.entries(keys)) {
        if (row[k] !== v) {
          match = false;
          break;
        }
      }
      if (match) return row[col];
    }
    return undefined;
  }

  /** Poll `pred(rows)` until it's true, or throw NotReady past `timeoutMs`. Registered as an
   * onChange listener internally, so it costs no polling interval of its own (no periodic
   * re-check, no lag or wasted work between checks) -- but it does NOT resolve any faster than
   * `onChange` itself fires: `waitFor` only sees what `onChange` decides to deliver, so it
   * inherits that emission's leading-edge/trailing-coalesce window (immediate if the table has
   * been quiet for `flushMs`, otherwise merged into the next scheduled flush) just like any other
   * listener. */
  waitFor(pred: (rows: readonly Row[]) => boolean, timeoutMs = 30_000): Promise<readonly Row[]> {
    if (pred(this.rowsFrozen)) return Promise.resolve(this.rowsFrozen);
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        unsubscribe();
        reject(new NotReady(`waitFor on '${this.tableName}' timed out after ${timeoutMs}ms`));
      }, timeoutMs);
      const unsubscribe = this.onChange((rows) => {
        if (pred(rows)) {
          clearTimeout(timer);
          unsubscribe();
          resolve(rows);
        }
      });
    });
  }

  /** Subscribe to change notifications; returns an unsubscribe function. Callbacks receive the
   * same frozen array `.rows` would return at that instant. */
  onChange(cb: ChangeListener): () => void {
    this.listeners.add(cb);
    return () => {
      this.listeners.delete(cb);
    };
  }

  [Symbol.asyncIterator](): AsyncIterator<readonly Row[]> {
    return {
      next: (): Promise<IteratorResult<readonly Row[]>> => {
        if (this.iterBuffer !== null) {
          const value = this.iterBuffer;
          this.iterBuffer = null;
          return Promise.resolve({ value, done: false });
        }
        if (this.iterDone) return Promise.resolve({ value: undefined, done: true });
        return new Promise((resolve) => this.iterWaiters.push(resolve));
      },
    };
  }

  close(): void {
    if (this.closed) return;
    this.closed = true;
    this.readerAbort.abort();
    // A pending trailing-coalesce timer must not fire after close (it would call listeners/touch
    // iterBuffer on a table nobody can observe results from anymore), and the touched keys it was
    // about to merge are not a resource that needs draining -- just discard both.
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    this.pendingTouched = null;
    this.iterDone = true;
    for (const w of this.iterWaiters.splice(0)) w({ value: undefined, done: true });
    for (const w of this.readyWaiters.splice(0)) {
      w.reject(new StreamsForgeError(`'${this.tableName}' was closed before it became ready`));
    }
  }

  /** Waits for the reader loop to actually unwind (subscription torn down, etc), not just the
   * `closed` flag being set -- useful in tests that want a clean process exit. */
  async closeAndWait(): Promise<void> {
    this.close();
    await this.readerLoopPromise;
  }

  async [Symbol.asyncDispose](): Promise<void> {
    await this.closeAndWait();
  }

  // ---- reader loop ----

  private async runReaderLoop(): Promise<void> {
    let backoff = 1000;
    let firstAttempt = true;
    while (!this.closed) {
      try {
        await this.subscribeSnapshotReplay(firstAttempt);
        backoff = 1000;
      } catch (err) {
        if (this.closed) return;
        this.readyFlag = false;
        this.reconnectsCount += 1;
        console.warn(
          `streamsforge: ${this.tableName} reader error (reconnect #${this.reconnectsCount} in ${backoff}ms): ${String(err)}`,
        );
        await sleep(backoff, this.readerAbort.signal);
        if (this.closed) return;
        backoff = Math.min(MAX_BACKOFF_MS, backoff * 2);
      }
      firstAttempt = false;
    }
  }

  private async subscribeSnapshotReplay(_firstAttempt: boolean): Promise<void> {
    // A resumed connection without a fresh snapshot silently corrupts the Z-set (deltas emitted
    // while it was down are gone), so every (re)connect starts from a clean reducer.
    this.zset = new ZSet(this.keyFields);

    // A fresh AbortController per attempt: closing an earlier attempt must never cancel a later
    // reconnect's subscription.
    this.readerAbort = new AbortController();
    const signal = this.readerAbort.signal;

    // MUST be awaited before the snapshot read is issued, not raced against it: Transport.subscribe()
    // only resolves once the subscription is actually registered with the server (see
    // transport.ts's doc comment). Racing the two here -- e.g. calling snapshot() first and only
    // then awaiting subscribe() -- would reopen the exact bug that contract exists to close: a
    // delta emitted after the server computes the snapshot but before it finishes registering us
    // is never sent at all (not buffered -- the server didn't know we existed yet), and a fresh
    // subscription gets no backfill to cover for it. Once this resolves, though, the subscription
    // and the snapshot read genuinely race each other (both can now proceed concurrently), which
    // is what the buffering below is for.
    const subscription = await this.transport.subscribe(this.tableName, signal);
    const iter = subscription[Symbol.asyncIterator]();

    // Race the initial snapshot read against incoming delta batches, buffering the latter --
    // this is the JS analogue of live.py's "drain the queue non-blocking, then seed, then
    // replay": no separate thread is needed because `Promise.race` between an in-flight
    // `snapshot()` and the next `iter.next()` naturally interleaves the two without ever
    // blocking the event loop.
    const buffered: Array<readonly [Delta[], number]> = [];
    const snapshotPromise = this.transport.snapshot(this.tableName);
    let pendingNext = iter.next();

    let snapshotResult: readonly [Delta[], number] | null = null;
    while (snapshotResult === null) {
      const winner = await Promise.race([
        snapshotPromise.then((v) => ({ kind: "snapshot" as const, v })),
        pendingNext.then((v) => ({ kind: "delta" as const, v })),
      ]);
      if (winner.kind === "snapshot") {
        snapshotResult = winner.v;
      } else {
        if (winner.v.done) {
          throw new StreamsForgeError(`'${this.tableName}' subscription ended before the initial snapshot`);
        }
        buffered.push(winner.v.value);
        pendingNext = iter.next();
      }
    }

    const [snapRows, snapSeq] = snapshotResult;
    this.zset.seed(snapRows);
    this.seqValue = snapSeq;
    for (const [deltas, seq] of buffered) {
      if (!this.zset.alreadyReflected(deltas)) {
        this.zset.apply(deltas);
        this.seqValue = seq;
      }
    }
    this.flushRowsSnapshot();

    if (this.closed) return;
    this.readyFlag = true;
    for (const w of this.readyWaiters.splice(0)) w.resolve();

    // BUG FIX: on a RECONNECT (not the first attempt), the reseed above just silently changed
    // `.rows` out from under every listener with no notification. `readyWaiters` is empty by now
    // regardless -- it was already drained on the first successful attempt, so the resolve loop
    // just above is a no-op here -- and nothing else calls emit() until the next LIVE delta
    // arrives, which on a quiet table can be never (this method's own top comment: deltas emitted
    // while the connection was down are gone, not buffered anywhere to replay later). So a
    // listener would keep showing pre-drop state indefinitely even though the table has already
    // silently moved on. Fix: emit right here, once, for every reconnect but the very first
    // connection (where nobody could be listening yet -- LiveTable.connect() hasn't returned a
    // handle to subscribe onChange() on, so emitting there would be a no-op at best and a
    // misleading "changed" notification for the initial state at worst -- hence `_firstAttempt`
    // gates it, guaranteeing exactly one emit per reconnect and none on first connect). `touched`
    // is every current key rather than a guess at what moved: everything may have changed across
    // the gap, and that is the honest signal to send.
    if (!_firstAttempt) {
      this.doEmit(new Set(this.zset.entries().map((e) => e.key)));
    }

    // Live loop: continue from `pendingNext`, which is already in flight and represents
    // whatever arrives next after the snapshot race settled.
    while (!this.closed) {
      const { value, done } = await pendingNext;
      if (done) {
        if (this.closed) return;
        throw new StreamsForgeError(`'${this.tableName}' subscription stream ended`);
      }
      const [deltas, seq] = value;
      const touched = this.zset.apply(deltas);
      this.seqValue = seq;
      this.scheduleEmit(touched);
      pendingNext = iter.next();
    }
  }

  private flushRowsSnapshot(): void {
    this.rowsFrozen = Object.freeze(this.zset.rows());
  }

  private scheduleEmit(touched: string[]): void {
    // ZSet.apply() reports every key whose presence or content actually changed (asserts,
    // retractions, and superseded stale keys alike -- see zset.ts's apply() doc comment), so an
    // empty `touched` here genuinely means this batch changed nothing: safe to skip the flush.
    if (touched.length === 0) return;
    if (this.pendingTouched) {
      for (const k of touched) this.pendingTouched.add(k);
      return; // a trailing flush is already scheduled; it will pick up this batch too
    }
    const now = Date.now();
    const elapsed = now - this.lastEmitAt;
    if (elapsed >= this.flushMs) {
      // Leading edge: at least `flushMs` has passed since the last emit (or there has never been
      // one), so tell the consumer right away -- no timer, no wait. This is exactly the case a
      // pure trailing window always got wrong: a lone update on an otherwise-quiet table. Also
      // covers `flushMs: 0` for free -- `elapsed >= 0` is always true, so every batch takes this
      // branch and nothing is ever coalesced.
      this.flushRowsSnapshot();
      this.doEmit(new Set(touched));
      return;
    }
    // Trailing coalesce: still inside the window opened by the last emit. Merge into ONE pending
    // emit fired at `lastEmitAt + flushMs` (i.e. `elapsed` less than a full `flushMs` from now,
    // not a fresh `flushMs` from now) -- a batch that lands late in the window must not push the
    // deadline back out.
    this.pendingTouched = new Set(touched);
    this.flushTimer = setTimeout(() => {
      const flushed = this.pendingTouched!;
      this.pendingTouched = null;
      this.flushTimer = null;
      this.flushRowsSnapshot();
      this.doEmit(flushed);
    }, this.flushMs - elapsed);
  }

  /** Every actual emission -- leading-edge immediate, trailing-coalesce timer, or the reconnect
   * fix's own emit -- funnels through here so `lastEmitAt` (what the leading-edge check measures
   * against) is always up to date. */
  private doEmit(touched: ReadonlySet<string>): void {
    this.lastEmitAt = Date.now();
    this.emit(touched);
  }

  private emit(touched: ReadonlySet<string>): void {
    const rows = this.rowsFrozen;
    for (const cb of this.listeners) {
      try {
        cb(rows, touched);
      } catch (err) {
        console.error(`streamsforge: onChange callback for '${this.tableName}' raised`, err);
      }
    }
    if (this.iterWaiters.length > 0) {
      for (const w of this.iterWaiters.splice(0)) w({ value: rows, done: false });
    } else {
      this.iterBuffer = rows; // latest-wins, capacity 1 -- see the field's own doc comment
    }
  }
}

File: clients/kotlin/src/main/kotlin/streamsforge/LiveTable.kt
package streamsforge
import java.io.Closeable

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.ReceiveChannel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.produceIn
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds
import kotlin.time.TimeSource

/**
 * One table's Z-set state, kept current by a coroutine (a child of the owning
 * [StreamsForgeClient]'s scope) that runs subscribe -> buffer -> snapshot -> replay -- see
 * [ZSet]'s docstring for why buffering is necessary -- and then keeps applying live deltas,
 * publishing through a LEADING edge + TRAILING coalesce window ([flushWindow], default 16ms --
 * one frame at 60Hz, the natural ceiling for a UI consumer that cannot display more than one
 * frame per 16ms anyway). If at least [flushWindow] has elapsed since the last publish, a batch
 * is published immediately after it's applied -- no delay, no wait -- so a lone update on an
 * otherwise-quiet table is never held back. Only a batch that lands INSIDE the window opened by
 * the previous publish gets merged into a single pending publish, fired at `lastPublish +
 * flushWindow`; further batches inside that same window merge into the same pending publish, so
 * at most one publish is ever pending. `flushWindow = ZERO` disables coalescing entirely and
 * publishes synchronously per applied batch. The window exists at all because a firehose of tens
 * of thousands of deltas/sec would otherwise fire one [StateFlow] emission per delta and melt a
 * collector, not because this repo's own hub-driven UI does anything similar --
 * `web/src/hooks/useTableRows.ts` publishes on every batch with no coalescing window of its own
 * (its 900ms timer is an unrelated flash-highlight effect).
 *
 * [rows] is an immutable snapshot, not a live view: built fresh from the reducer on every
 * publish, same reasoning as the Python client's `.df` -- a row list that mutated in place would
 * race the reducer with no change notification a consumer could rely on. [rowsFlow] is the same
 * thing as a [StateFlow] for consumers that want to react rather than poll -- a [StateFlow]
 * inherently conflates (only the latest value is ever delivered, assigning `.value` never
 * suspends), which is exactly the "latest-wins, not a queue" semantics a state-snapshot stream
 * needs: a slow collector never blocks the reader coroutine, and there is no backlog to grow.
 *
 * Closing a `LiveTable` cancels its own subscription only; closing the owning client cancels
 * every `LiveTable` it created (structured concurrency: this table's job is a child of the
 * client's).
 */
class LiveTable internal constructor(
    private val transport: TableTransport,
    private val tableName: String,
    private val keyFields: List<String>?,
    parentScope: CoroutineScope,
    private val flushWindow: Duration = DEFAULT_FLUSH_WINDOW,
) : Closeable {
    private val job = SupervisorJob(parentScope.coroutineContext[Job])
    private val scope = CoroutineScope(parentScope.coroutineContext + job)

    // Mutated only by the single reader coroutine below; external reads go through `_rows`
    // (StateFlow, safe to read cross-thread) rather than touching the reducer directly.
    private var zset = ZSet(keyFields)

    // Also touched only by the reader coroutine -- when the most recent publish happened, so the
    // leading-edge check in liveLoop() can tell whether the window has already elapsed. Null
    // means "no publish yet this session", which the check below treats as elapsed (the snapshot
    // fill in doSnapshotAndReplay sets this before liveLoop ever runs, so in practice this is only
    // ever null for the instant between construction and the first snapshot).
    private var lastPublish: TimeSource.Monotonic.ValueTimeMark? = null

    private val _rows = MutableStateFlow<List<Row>>(emptyList())
    val rowsFlow: StateFlow<List<Row>> = _rows.asStateFlow()
    val rows: List<Row> get() = _rows.value

    @Volatile var ready: Boolean = false
        private set

    @Volatile var reconnects: Int = 0
        private set

    @Volatile var seq: Long = 0
        private set

    private val readyGate = CompletableDeferred<Unit>()

    init {
        scope.launch { runLoop() }
    }

    /** Suspends until the first snapshot+replay lands, or [timeout] elapses -- called right
     * after construction by [StreamsForgeClient.table] (a coroutine can't block inside `init`,
     * unlike the Python client's constructor-time thread join). */
    internal suspend fun awaitReady(timeout: Duration) {
        val completed = withTimeoutOrNull(timeout) { readyGate.await() }
        if (completed == null) {
            close()
            throw NotReadyException(
                "table '$tableName' did not fill within $timeout -- a brand-new table gets no " +
                    "backfill, so this is expected until something pushes to it"
            )
        }
    }

    fun value(col: String, keys: Map<String, Any?> = emptyMap()): Any? =
        rows.firstOrNull { row -> keys.all { (k, v) -> row[k] == v } }?.get(col)

    /** Polls [pred] against [rows] until it returns true, or raises [NotReadyException] after
     * [timeout]. A predicate that throws (e.g. indexing a column that doesn't exist yet on an
     * empty table) is treated as "not yet", same as the Python client's `wait_for`. */
    suspend fun waitFor(timeout: Duration = 30.seconds, pred: (List<Row>) -> Boolean): List<Row> {
        val deadline = TimeSource.Monotonic.markNow() + timeout
        while (true) {
            val snapshot = rows
            if (runCatching { pred(snapshot) }.getOrDefault(false)) return snapshot
            if (deadline.hasPassedNow()) throw NotReadyException("waitFor on '$tableName' timed out after $timeout")
            delay(50)
        }
    }

    override fun close() {
        job.cancel()
    }

    // ---- reader coroutine ----

    private suspend fun runLoop() {
        var backoff = 1.seconds
        while (true) {
            try {
                subscribeSnapshotReplay()
                backoff = 1.seconds
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                ready = false
                reconnects++
                delay(backoff)
                backoff = minOf(backoff * 2, 15.seconds)
            }
        }
    }

    private suspend fun subscribeSnapshotReplay(): Unit = coroutineScope {
        // A resumed connection without a fresh snapshot silently corrupts the Z-set (deltas
        // emitted while it was down are gone) -- every (re)connect starts from a clean reducer.
        zset = ZSet(keyFields)
        // `subscribeTable` is itself suspend -- it returns only once the subscription is
        // registered (or as near to it as the transport can guarantee), which is what makes it
        // safe to call `snapshot()` right after without racing the live channel's own setup (see
        // [TableTransport.subscribeTable]'s doc).
        val channel = transport.subscribeTable(tableName).produceIn(this)
        try {
            doSnapshotAndReplay(channel)
            liveLoop(channel)
        } finally {
            channel.cancel()
        }
    }

    private suspend fun doSnapshotAndReplay(channel: ReceiveChannel<DeltaBatch>) {
        val (snapRows, snapSeq) = transport.snapshot(tableName)
        val buffered = drainNowait(channel) // arrived while the snapshot read was in flight

        zset.seed(snapRows)
        seq = snapSeq
        for (batch in buffered) {
            if (!zset.alreadyReflected(batch.deltas)) {
                zset.apply(batch.deltas)
                seq = batch.seq
            }
        }
        publishNow()
        ready = true
        readyGate.complete(Unit) // no-op on reconnect; only the first fill unblocks awaitReady()
    }

    private suspend fun liveLoop(channel: ReceiveChannel<DeltaBatch>) {
        while (true) {
            val first = channel.receive() // throws ClosedReceiveChannelException when the stream ends
            zset.apply(first.deltas)
            seq = first.seq

            if (flushWindow == Duration.ZERO) {
                publishNow() // coalescing disabled -- publish this batch on its own
                continue
            }

            val last = lastPublish
            if (last == null || last.elapsedNow() >= flushWindow) {
                // LEADING EDGE: the window has already elapsed since the last publish (or there
                // has never been one), so this batch -- lone or not -- goes out with no delay.
                publishNow()
                continue
            }

            // TRAILING COALESCE: still inside the window opened by the last publish. Keep
            // draining the channel (never stalling it) until that window closes, folding
            // whatever else arrives into the same reducer state, then publish exactly once.
            val deadline = last + flushWindow
            while (true) {
                val remaining = deadline - TimeSource.Monotonic.markNow()
                if (remaining <= Duration.ZERO) break
                val next = withTimeoutOrNull(remaining) { channel.receive() } ?: break
                zset.apply(next.deltas)
                seq = next.seq
            }
            publishNow()
        }
    }

    /** The only place `_rows` is assigned -- also stamps [lastPublish] so the leading-edge check
     * above measures from the true last publish, not from whenever a batch happened to arrive. */
    private fun publishNow() {
        _rows.value = zset.rows()
        lastPublish = TimeSource.Monotonic.markNow()
    }

    private fun drainNowait(channel: ReceiveChannel<DeltaBatch>): List<DeltaBatch> {
        val items = mutableListOf<DeltaBatch>()
        while (true) {
            val result = channel.tryReceive()
            if (result.isSuccess) items.add(result.getOrThrow()) else break
        }
        return items
    }

    companion object {
        /** One frame at 60Hz -- a UI cannot display more than one frame per 16ms, so it's the
         * natural ceiling for a change-notification window rather than a compromise. */
        val DEFAULT_FLUSH_WINDOW = 16.milliseconds
    }
}
